wolfcrypt/src/random.c

```diff
@@ -800,6 +809,777 @@ int wc_RNG_TestSeed(const byte* seed, word32 seedSz)
 #endif /* HAVE_HASHDRBG */
 /* End NIST DRBG Code */
 
+#ifdef HAVE_ENTROPY_MEMUSE
+/* Define ENTROPY_MEMUSE_THREAD to force use of counter in a new thread.
+ * Only do this when high resolution timer not otherwise available.
+ */
+
+/* Number of bytes that will hold the maximum entropy bits. */
+#define MAX_ENTROPY_BYTES    (MAX_ENTROPY_BITS / 8)
+/* Number of bits stored for one sample. */
+#define ENTROPY_BITS_USED    8
+
+/* Minimum entropy from a sample. */
+#define ENTROPY_MIN          1
+/* Number of extra samples to ensure full entropy. */
+#define ENTROPY_EXTRA        64
+/* Maximum number of bytes to sample to produce max entropy. */
+#define MAX_NOISE_CNT        (MAX_ENTROPY_BITS * 8 + ENTROPY_EXTRA)
+
+/* MemUse entropy global state initialized. */
+static int entropy_memuse_initialized = 0;
+/* Global SHA-3 object used for conditioning entropy and creating noise. */
+static wc_Sha3 entropyHash;
+/* Reset the health tests. */
+static void Entropy_HealthTest_Reset(void);
+
+#if !defined(ENTROPY_MEMUSE_THREAD) && \
+    (defined(__x86_64__) || defined(__i386__))
+/* Get the high resolution time counter.
+ *
+ * @return  64-bit count of CPU cycles.
+ */
+static WC_INLINE word64 Entropy_TimeHiRes(void)
+{
+    unsigned int lo_c, hi_c;
+    __asm__ __volatile__ (
+        "rdtsc"
+            : "=a"(lo_c), "=d"(hi_c)   /* out */
+            : "a"(0)                   /* in */
+            : "%ebx", "%ecx");         /* clobber */
```

Comment:
I asked that question about `--no-asm` because this popped out at me.

Reply:
This is for the high resolution timer and is in the `__x86_64__` and `__i386__` section. It all build fine with `--disable-asm` with or without threading. You are welcome to review the code more closely and test since you seem so interested in it. I am likely to merge this any minute.

Reply:
@anhu That popped out for me too.
Turned out `--disable-asm `doesn't turn off inline assembly.
inline assembly gets built with both `--enable-asm` (Default) or `--disable-asm`

However, if you configure with `./configure --enable-entropy-memuse --enable-intelasm` then we default to using `rdseed`, hardware generated random value instead of memuse_entropy